text/0000-proc-macro-expansion-api.md

```diff
@@ -0,0 +1,411 @@
+
+This doesn't just stop at variable names. The above principles apply to mods, structs, trait definition, trait method calls, macros - anything with a name which needs to be looked up.
+
+### Unhygienic Scopes
+
+Importantly, macro hygiene is _optional_: since we can manipulate the spans on tokens, we can change how a variable is resolved. For example:
+
+```rust
+extern crate proc_macro;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::Span;
+
+#[proc_macro]
+fn my_unhygienic_macro(tokens: TokenStream) -> TokenStream {
+    let hygienic = quote_spanned! { Span::def_site(),
+        let mut x = 0; // [Def]
+    };
+    let unhygienic = quote_spanned! { Span::call_site(),
+        x += 1;        // [Call]
+    };
+    quote! {
+        #hygienic      // [Def]
+        #tokens        // [Call]
+        #unhygienic    // [Call]
+    }
+}
+```
+
+If we call `my_unhygienic_macro` instead of `my_hygienic_macro` in `main` as before, the result is:
+
+```rust
+fn main() {
+    let mut x = 0; // 1. [Def]
+    let mut x = 1; // 2. [Call], from main
+    x += 2;        // 3. [Call], from main
+    x += 1;        // 4. [Call], from my_unhygienic_macro
+    println!(x);   // 5. [Call]
+}
+```
+
+By changing the scope of the span of the tokens on line 4 (using `quote_spanned` instead of `quote`), that instance of `x` will resolve to the one defined on line 2 instead of line 1. In fact, the variable actually declared by our macro on line 1 is never used.
+
+This trick has a few uses, such as 'exporting' a name to the caller of the macro. If hygiene was not optional, any new functions or modules you created in a macro would only be resolvable in the same macro.
+
+There are also some interesting [examples](https://github.com/dtolnay/syn/blob/030787c71b4cfb2764bccbbd2bf0e8d8497d46ef/examples/heapsize2/heapsize_derive/src/lib.rs#L65) of how this gets used to resolve method calls on traits declared in `[Def]`, but called with variables from `[Call]`.
+
+## API Overview
+
+The full API provided by `proc_macro` and used by `syn` is more flexible than suggested by the use of `parse_expand` and `parse_meta_expand` above. To begin, `proc_macro` defines a struct, `MacroCall`, with the following interface:
+
+```rust
+struct MacroCall {...};
+
+impl MacroCall {
+    fn new_proc(path: TokenStream, args: TokenStream) -> Self;
+    
+    fn new_attr(path: TokenStream, args: TokenStream, body: TokenStream) -> Self;
+    
+    fn call_from(self, from: Span) -> Self;
+    
+    fn expand(self) -> Result<TokenStream, Diagnostic>;
+}
+```
+
+The functions `new_proc` and `new_attr` create a procedural macro call and an attribute macro call, respectively. Both expect `path` to parse as a [path](https://docs.rs/syn/0.12/syn/struct.Path.html) like `println` or `::std::println`. The scope of the spans of `path` are used to resolve the macro definition. This is unlikely to work unless all the tokens have the same scope.
+
```

Comment:
Overall, I really like the idea of using a Builder API - it keeps things simple and is future-proof